File: front-end/src/app/domain/models/game/card/impl/JailBreakCard.ts
import {Card, CardType} from "../Card";
import {GameAsset} from "../../GameAsset";

export class JailBreakCard extends Card implements GameAsset {
  private _initialPurchasePrice: number;

  public constructor(
    id: number,
    description: string,
    type: CardType,
    initialPurchasePrice: number
  ) {
    super(JailBreakCard.name, id, description, type);
    this._initialPurchasePrice = initialPurchasePrice;
  }

  public static trueCopy(action: JailBreakCard): JailBreakCard | null {
    return new JailBreakCard(
      action.id,
      action.description,
      action.type,
      action.initialPurchasePrice
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(),
      {
        "initialPurchasePrice": this.initialPurchasePrice
      }
    );
  }

  public get initialPurchasePrice(): number {
    return this._initialPurchasePrice;
  }

  public override toString(): string {
    return "Get-out-of-jail card";
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/card/impl/JailBreakCard.java
package nl.hva.backend.domain.models.game.card.impl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.GameAsset;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.card.Card;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.ManyToOne;

/**
 * TODO JavaDoc
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class JailBreakCard extends Card implements GameAsset {
    private double initialPurchasePrice;

    /**
     * TODO JavaDoc
     */
    public JailBreakCard(
            String description,
            CardType type,
            double initialPurchasePrice
    ) {
        super(description, type);
        this.initialPurchasePrice = initialPurchasePrice;
    }


    /**
     * TODO JavaDoc
     *
     * @author Hamza el Haouti
     */
    @Entity
    @EntityListeners({AuditingEntityListener.class})
    @Getter
    @Setter
    @NoArgsConstructor
    public static class JailBreakCardBid extends GameAssetBid {

        @ManyToOne
        private JailBreakCard card;

        /**
         * TODO JavaDoc
         */
        public JailBreakCardBid(
                Player owner,
                Player bidder,
                double amount,
                JailBreakCard card
        ) {
            super(owner, bidder, amount);
            this.card = card;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof JailBreakCardBid that)) return false;
            return getId() == that.getId();
        }

        @Override
        public int hashCode() {
            return (int) (getId() ^ (getId() >>> 32));
        }
    }

}